src/transformers/modeling_utils.py

```diff
@@ -1151,6 +1156,28 @@ def __init__(self, config: PretrainedConfig, *inputs, **kwargs):
         self.warnings_issued = {}
         self.generation_config = GenerationConfig.from_model_config(config) if self.can_generate() else None
 
+        # TODO: This is TEMPORARY and need to be discussed, should it rather be in XXXPreTrainedModel __init__?
+        if config.attn_implementation == "flash_attention_2":
+            if not self._supports_flash_attn_2:
+                raise ValueError(
+                    f'Passed config.attn_implementation == "flash_attention_2" but {self.__class__.__name__} does not support Flash Attention yet.'
+                )
+
+            if not is_flash_attn_2_available():
+                raise ImportError(
+                    "Flash Attention 2 is not available. Please refer to the documentation of https://github.com/Dao-AILab/flash-attention for"
+                    " installing it. Make sure to have at least the version 2.1.0"
+                )
+
+        if config.attn_implementation == "sdpa":
+            if not self._supports_sdpa:
+                raise ValueError(
+                    f'Passed config.attn_implementation == "sdpa" but {self.__class__.__name__} does not support SDPA yet.'
+                )
+
+            if not is_torch_sdpa_available():
+                raise ImportError("SDPA is not available. Please use torch>=2.1.1 in order to use SDPA.")
+
```

Comment:
(got outdated so reposting)

Please give your opinion on this.

The context is the following: the current dispatch between normal attention and flash attention (e.g. https://github.com/huggingface/transformers/blob/638d49983f36af910934b38771b4e55c835c1774/src/transformers/models/llama/modeling_llama.py#L629-L633) fails to check FA2 availability if the model is initialized from a `config` rather than from `from_pretrained`. Beside, this kind of check happening at the layer level is unnecessary and may be done only once.

By moving the `is_flash_attn_2_available()` and `is_torch_sdpa_available()` checks directly in `XXXXPreTrainedModel` or `PreTrainedModel` init, we avoid many checks in the init that are moved to a higer level.

Do you agree? Should this rather be in `XXXXPreTrainedModel` init?

Reply:
```py
 self.self_attn = ( 
     LlamaAttention(config=config) 
     if not getattr(config, "_flash_attn_2_enabled", False) 
     else LlamaFlashAttention2(config=config) 
 )
```

I think we should refactor this to also only rely on the model and not the model's config - cc @younesbelkada wdyt?

Reply:
I disagree, let's move the discussion to the related thread below https://github.com/huggingface/transformers/pull/26572#discussion_r1402354316. The piece of code I mention here got refactored so resolving this thread.